src/openrct2/scripting/ScriptEngine.cpp

```diff
@@ -1767,6 +1767,50 @@ std::string OpenRCT2::Scripting::Stringify(const DukValue& val)
     return ExpressionStringifier::StringifyExpression(val);
 }
 
+ColourWithFlags OpenRCT2::Scripting::processInputWindowColour(const DukValue& value, int32_t apiVersion)
+{
+    auto c = ColourWithFlags{ COLOUR_BLACK };
+    if (value.type() == DukValue::Type::NUMBER)
+    {
+        auto number = value.as_int();
+        if (apiVersion < API_VERSION_89_WINDOW_COLOURS)
+            c = ColourWithFlags::fromLegacy(number);
+        else
+            c.colour = number;
+    }
+    else if (value.type() == DukValue::Type::OBJECT)
+    {
+        if (value["colour"].type() == DukValue::Type::STRING)
+            c.colour = Colour::FromString(value["colour"].as_string(), COLOUR_BLACK);
+        else
+            c.colour = value["colour"].as_int();
+
+        if (value["translucent"].as_bool())
+            c.flags |= EnumToFlag(ColourFlag::translucent);
+    }
+
+    // Normalise all values
+    c.colour = std::clamp<uint16_t>(c.colour, COLOUR_BLACK, COLOUR_COUNT - 1);
+    c.flags &= EnumToFlag(ColourFlag::translucent);
+
+    return c;
+}
+
+DukValue OpenRCT2::Scripting::processOutputWindowColour(duk_context& ctx, const ColourWithFlags& value, int32_t apiVersion)
+{
+    if (apiVersion < API_VERSION_89_WINDOW_COLOURS)
+    {
+        return ToDuk(&ctx, value.colour);
+    }
+    else
+    {
+        DukObject result(&ctx);
+        result.Set("colour", Colour::ToString(value.colour));
+        result.Set("translucent", value.hasFlag(ColourFlag::translucent));
```

Comment:
Can we possibly add a third property for a `id: number` for the colour number? During testing I found some places where the new string is not compatible with API's still using the number and currently there's not really a way to deal with that except for copying the entire table into the Javascript plugin.

Reply:
Plugins can only read from this data, right? If so, then yes.